File: android/src/main/java/com/visioncamerafacedetector/VisionCameraFaceDetectorPluginPackage.java
package com.visioncamerafacedetector;

import androidx.annotation.NonNull;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;
import com.mrousavy.camera.frameprocessor.FrameProcessorPlugin;
import com.mrousavy.camera.frameprocessor.FrameProcessorPluginRegistry;
import com.visioncamerafacedetector.VisionCameraFaceDetectorPlugin;

import java.util.Collections;
import java.util.List;

public class VisionCameraFaceDetectorPluginPackage implements ReactPackage {
  static {
    FrameProcessorPluginRegistry.addFrameProcessorPlugin("detectFaces", options -> new VisionCameraFaceDetectorPlugin(options));
  }

  @NonNull
  @Override
  public List<NativeModule> createNativeModules(@NonNull ReactApplicationContext reactContext) {
    return Collections.emptyList();
  }

  @NonNull
  @Override
  public List<ViewManager> createViewManagers(@NonNull ReactApplicationContext reactContext) {
    return Collections.emptyList();
  }
}

File: android/src/main/java/com/visioncamerafacedetector/VisionCameraFaceDetectorPluginPackage.kt
package com.visioncamerafacedetector

import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.uimanager.ViewManager
import com.mrousavy.camera.frameprocessors.FrameProcessorPluginRegistry

class VisionCameraFaceDetectorPluginPackage: ReactPackage {
  companion object {
    private var orientationManager: VisionCameraFaceDetectorOrientation? = null

    init {
      FrameProcessorPluginRegistry.addFrameProcessorPlugin("detectFaces") { proxy, options ->
        if(orientationManager == null) {
          orientationManager = VisionCameraFaceDetectorOrientation(proxy.context)
        }
        VisionCameraFaceDetectorPlugin(options, orientationManager!!)
      }
    }

    fun stopDeviceOrientationListener() {
      orientationManager?.stopDeviceOrientationListener()
      orientationManager = null
    }
  }

  override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
    return listOf(VisionCameraFaceDetectorOrientationManager(reactContext))
  }


  override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
    return emptyList()
  }
}

class VisionCameraFaceDetectorOrientationManager(context: ReactApplicationContext) :
  ReactContextBaseJavaModule(context) {
  override fun getName(): String {
    return "VisionCameraFaceDetectorOrientationManager"
  }

  @ReactMethod
  fun stopDeviceOrientationListener() {
    VisionCameraFaceDetectorPluginPackage.stopDeviceOrientationListener()
  }
}
